### ISSUE TITLE
Weird lifetime error on closure in thread::scope

### ISSUE THREAD

**Issue by m-ou-se** [rust-lang/rust#95527]
This was reported here: https://github.com/rust-lang/rust/issues/93203#issuecomment-1084630290

This should compile, but this fails to compile:

```rust
#![feature(scoped_threads)]

pub fn foo(x: &u8) {
    std::thread::scope(|s| {
        s.spawn(|| drop(x));
    });
}
```

Error:

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/main.rs:5:17
  |
5 |         s.spawn(|| drop(x));
  |                 ^^^^^^^^^^
  |
note: first, the lifetime cannot outlive the anonymous lifetime defined here...
 --> src/main.rs:3:15
  |
3 | pub fn foo(x: &u8) {
  |               ^^^
note: ...so that the types are compatible
 --> src/main.rs:5:17
  |
5 |         s.spawn(|| drop(x));
  |                 ^^^^^^^^^^
  = note: expected `(&&u8,)`
             found `(&&u8,)`
note: but, the lifetime must be valid for the anonymous lifetime #1 defined here...
 --> src/main.rs:4:24
  |
4 |       std::thread::scope(|s| {
  |  ________________________^
5 | |         s.spawn(|| drop(x));
6 | |     });
  | |_____^
note: ...so that the types are compatible
 --> src/main.rs:5:11
  |
5 |         s.spawn(|| drop(x));
  |           ^^^^^
  = note: expected `&Scope<'_, '_>`
             found `&Scope<'_, '_>`
```

(rustc 1.61.0-nightly (1eb72580d 2022-03-08))

It seems to be complaining that it cannot find a lifetime no longer than the one of the `x: &u8` argument, but no shorter than the scope closure. The error is very hard to read though.

It's interesting how the capture tuple `(&&u8,)` of the closure is visible in the diagnostic.

**Comment by m-ou-se** [comment]
It compiles fine with `#![feature(nll)]` enabled.

**Comment by danielhenrymantilla** [comment]
So, I think I've managed to reduce it a bit further:

```rust
fn bar<'at_x, 'x> (at_x: &'at_x &'x u8)
{
    ::std::thread::scope::<'at_x>(move |s| {
        s.spawn(move || drop(*at_x));
    });
}
```

passes, but if we remove the `'at_x` annotation (the value of `'env`):

```rs
fn bar<'at_x, 'x> (at_x: &'at_x &'x u8)
{
    ::std::thread::scope::<'_>(move |s| {
        s.spawn(move || drop(*at_x));
    });
}
```

**Comment by krdln** [comment]
I don't get why `#![feature(nll)]` makes it compile – wasn't the whole idea of currently enabled "migration mode" to run old borrowck and nll just for cases nll rejects, but old borrowck accepts and thus we issue compat warning? I'm very confused... Because if old borrowck still have a veto posibility, then I don't get how even we can accept "problem case #1" and "problem case #2" now (which we do)

**Comment by m-ou-se** [comment]
`library/std/src/collections/hash/map.rs` also has code that doesn't compile if we remove `#![feature(nll)]`.

**Comment by bjorn3** [comment]
Should this be closed now that https://github.com/rust-lang/rust/pull/95565 has been merged?
